File: mobile/src/api/live.ts
/**
 * Hooks for the values the scraper stores from Open-Meteo (the CAMS air-quality model). Types
 * mirror backend/app/schemas/live.py.
 *
 * The values are model estimates for the ~45 km area around a city, not station measurements, and
 * the AQI is CPCB's formula applied to them, so the cards always say so and show the source credit
 * that the data's licence asks for.
 */

import { useQuery } from '@tanstack/react-query';

import { apiGet } from './client';
import type { AqiCategoryKey, City, CityGroup } from './types';

export type Pollutant = 'PM2.5' | 'PM10' | 'NO2' | 'SO2' | 'CO' | 'O3' | 'NH3';

export interface LiveReading {
  pollutant: Pollutant;
  value: number;
  unit: string;
  observed_at: string;
}

export interface EstimatedAqi {
  value: number;
  category: AqiCategoryKey;
  dominant: Pollutant;
  sub_indices: Partial<Record<Pollutant, number>>;
}

/** Where live values come from; `attribution` must be shown, linked, next to them. */
export interface LiveSource {
  source: string;
  attribution: string;
  attribution_url: string;
  stale_after_hours: number;
}

/** GET /v1/live/cities/{id} */
export interface CityLive extends LiveSource {
  city: City;
  status: 'ok' | 'stale' | 'no_recent_readings' | 'not_in_feed';
  measure: 'model_estimate';
  observed_at: string | null;
  fetched_at: string | null;
  aqi: EstimatedAqi | null;
  readings: LiveReading[];
}

/** GET /v1/live/cities */
export interface LiveCitiesResponse extends LiveSource {
  cities: { city: City; aqi: EstimatedAqi; observed_at: string }[];
}

// The scraper stores a new hour once an hour, so refetching more often than this is wasted work.
const LIVE_STALE_TIME_MS = 5 * 60 * 1000;

export function useLiveCities(params: {
  state?: string | null;
  group?: CityGroup | null;
  order?: 'asc' | 'desc';
  limit?: number;
}) {
  return useQuery({
    queryKey: ['live', 'cities', params],
    queryFn: () => apiGet<LiveCitiesResponse>('/v1/live/cities', params),
    staleTime: LIVE_STALE_TIME_MS,
  });
}

export function useCityLive(cityId: number | null) {
  return useQuery({
    queryKey: ['live', 'city', cityId],
    queryFn: () => apiGet<CityLive>(`/v1/live/cities/${cityId}`),
    enabled: cityId !== null,
    staleTime: LIVE_STALE_TIME_MS,
    // The screen stays open while new hours arrive; check again every quarter hour.
    refetchInterval: 15 * 60 * 1000,
  });
}

File: backend/app/schemas/live.py
"""Response models for the live endpoints: the scraper's model values and its runs."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

from app.domain import AqiCategory, Pollutant
from app.schemas.history import CityOut


class LiveReadingOut(BaseModel):
    pollutant: Pollutant
    value: float
    unit: str
    observed_at: datetime


class EstimatedAqiOut(BaseModel):
    """CPCB's AQI formula applied to the model's concentrations: an estimate, not an official
    AQI."""

    value: int
    category: AqiCategory
    dominant: Pollutant
    sub_indices: dict[Pollutant, int]


class LiveSourceOut(BaseModel):
    """Where live values come from. Open-Meteo's licence asks apps to show the attribution, with
    a link, next to the data."""

    source: str
    attribution: str
    attribution_url: str
    stale_after_hours: int


class CityLiveOut(LiveSourceOut):
    """The newest model values for one city.

    `status` says what the app can show: ok; stale (the newest hour is older than
    `stale_after_hours`); no_recent_readings (nothing in the last 48 hours); or not_in_feed (the
    city has no coordinates, so the scraper can't read the model there).
    """

    city: CityOut
    status: Literal["ok", "stale", "no_recent_readings", "not_in_feed"]
    # Concentrations from a model of the ~45 km area around the city, not a station measurement.
    measure: Literal["model_estimate"] = "model_estimate"
    observed_at: datetime | None  # the newest hour, IST
    fetched_at: datetime | None
    aqi: EstimatedAqiOut | None  # None when CPCB's minimum-data rules aren't met
    readings: list[LiveReadingOut]


class LiveCityRankOut(BaseModel):
    city: CityOut
    aqi: EstimatedAqiOut
    observed_at: datetime


class LiveCitiesResponse(LiveSourceOut):
    cities: list[LiveCityRankOut]


class ScrapeRunOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    source: str
    status: str
    started_at: datetime
    finished_at: datetime | None
    cities_requested: int
    cities_seen: int
    records_seen: int
    records_skipped: int
    readings_inserted: int
    readings_purged: int
    error: str | None


class LiveStatus(BaseModel):
    last_run: ScrapeRunOut | None
    last_success: ScrapeRunOut | None
    data_as_of: datetime | None
    cities_with_readings: int  # in the last 48 hours
